pkg/logger/logger.go

```diff
@@ -2,48 +2,94 @@ package logger
 
 import (
 	"fmt"
+	"os"
 	"strings"
 
 	"github.com/go-logr/logr"
+	"github.com/go-logr/zapr"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 	ctrlzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
 
 	"github.com/dominodatalab/hephaestus/pkg/config"
 )
 
+var (
+	consoleEncoder zapcore.Encoder
+	jsonEncoder    zapcore.Encoder
+)
+
 func New(cfg config.Logging) (logr.Logger, error) {
-	opts := []ctrlzap.Opts{
-		ctrlzap.UseDevMode(cfg.Development),
-		func(opts *ctrlzap.Options) { opts.TimeEncoder = zapcore.RFC3339NanoTimeEncoder },
+	// container logging
+	var containerEncoder zapcore.Encoder
+	enc := strings.ToLower(cfg.Container.Encoder)
+
+	if enc == "" || enc == "console" {
+		containerEncoder = consoleEncoder
+	} else if enc == "json" {
+		containerEncoder = jsonEncoder
+	} else {
+		return logr.Logger{}, fmt.Errorf("%q is an invalid encoder", enc)
 	}
 
-	if cfg.Encoder != "" {
-		switch strings.ToLower(cfg.Encoder) {
-		case "console":
-			opts = append(opts, ctrlzap.ConsoleEncoder())
-		case "json":
-			opts = append(opts, ctrlzap.JSONEncoder())
-		default:
-			return logr.Logger{}, fmt.Errorf("invalid logger encoder %q", cfg.Encoder)
-		}
+	ll, err := parseLevel(cfg.Container.LogLevel)
+	if err != nil {
+		return logr.Logger{}, fmt.Errorf("invalid container log level: %w", err)
 	}
 
-	if cfg.LogLevel != "" {
-		lvl := zap.NewAtomicLevel()
-		if err := lvl.UnmarshalText([]byte(cfg.LogLevel)); err != nil {
-			return logr.Logger{}, fmt.Errorf("cannot set logger level: %w", err)
-		}
-		opts = append(opts, ctrlzap.Level(lvl))
+	cores := []zapcore.Core{
+		zapcore.NewCore(&ctrlzap.KubeAwareEncoder{Encoder: containerEncoder}, zapcore.Lock(os.Stdout), ll),
 	}
 
-	if cfg.StacktraceLevel != "" {
-		lvl := zap.NewAtomicLevel()
-		if err := lvl.UnmarshalText([]byte(cfg.StacktraceLevel)); err != nil {
-			return logr.Logger{}, fmt.Errorf("cannot set logger stacktrace level: %w", err)
+	// logfile logging
+	if cfg.Logfile.Enabled {
+		file, err := os.Create(cfg.Logfile.Filepath)
+		if err != nil {
+			return logr.Logger{}, fmt.Errorf("cannot create logfile logger: %w", err)
 		}
-		opts = append(opts, ctrlzap.StacktraceLevel(lvl))
+
+		level, err := parseLevel(cfg.Logfile.LogLevel)
+		if err != nil {
+			return logr.Logger{}, fmt.Errorf("invalid logfile log level: %w", err)
+		}
+		fileCore := zapcore.NewCore(&ctrlzap.KubeAwareEncoder{Encoder: jsonEncoder}, zapcore.Lock(file), level)
```

Comment:
Do we need to consider [file rotation](https://github.com/uber-go/zap/blob/master/FAQ.md#does-zap-support-log-rotation)? Feel like we chatted about it, but I don't recall what the outcome was

Reply:
This totally slipped my mind. I think we _should_ control the logfile size but I'll need to test what, if any, extra configuration is required by Vector to make sure it's not stuck on an old file inode.

Reply:
It looks like the default `file` source from Vector automatically picks up a file rotation event and starts tailing the new logfile.

```
hephaestus-manager-58dfdff677-pkmtq log-processor 2022-03-30T16:59:18.907340Z  INFO source{component_kind="source" component_id=hephaestus_logs component_type=f
ile component_name=hephaestus_logs}:file_server: vector::internal_events::file::source: Found new file to watch. file=/var/log/hephaestus/output.json
```